# Exploratory analysis - Event data

The goal of this notebook is to give you a quick overview over the data that is used in the following notebooks. In this notebook you will:
- Learn how to read in the different data sources we are going to use
- Get a first understanding of the data
- Get to know a great Python package called pandas_profiling

In total there are 5 different data sources that we are going to deal with:
- [Player data](#player_data)
- [Team data](#team_data)
- [Match data](#match_data)
- [Formation data](#formation_data) 
- [Event data](#event_data)

We will look into them on a relatively high level as we are going to go into more detail in the following notebooks. So don't worry if you have the feeling that you have not completely understood the data in depth after this notebook. The only important thing is, that you know which data sources there are and which information they contain.

I already have to tell you that this notebook will not be the most exciting one, but it will be rather dry. :-( On purpose, I do not yet go into a lot of visualization, as we will do this in the following notebooks. However,I promise that the other notebooks will be more entertaining! :-) 

Notice that we will be using data that is already preprocessed and uploaded to GitHub. In case you want to know more about the data source and how to do the preprocessing yourself, you can check out [this](https://figshare.com/collections/Soccer_match_event_dataset/4415000/2) website as well as the code in the [preprocessing folder](https://github.com/CleKraus/soccer_analytics/tree/master/preprocessing/cleaning_wyscout_data).

In [1]:
import os
import pandas as pd
from pandas_profiling import ProfileReport

if os.getcwd().split(os.sep)[-1] == "notebooks":
    os.chdir("../")
    
import helper.io as io

import importlib
importlib.reload(io)

# this is very useful as it makes sure that always all columns and rows of a data frame are displayed
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

<a id="player_data"></a>

## Player data

You can use the *read_xxx_data* functions from *helper.io* to read in all the data sources mentioned above. Let's start by looking into the player data.

In [2]:
df_players = io.read_player_data()

In [3]:
df_players.head()

playerId playerStrongFoot   playerName playerPosition
0     32777            right     H. Tekin             GK
1    393228             left      M. Sarr             DF
2    393230          unknown  O. Mandanda             GK
3     32793            right   A. N'Diaye             MD
4    393247            right    I. Konaté             DF

I think all of the columns are self-explaining. Let's quickly check how many players we have in total...

In [4]:
print(f"Total numbers of players: {len(df_players)}")

Total numbers of players: 3603


... and let's make sure that we do not have any duplicates in the player Id:

In [5]:
print(f"Number of unique playerIds: {len(df_players['playerId'].unique())}")

Number of unique playerIds: 3603


What about the number of players we have for each position?

In [6]:
df_players.groupby("playerPosition").size().sort_values()

playerPosition
GK     426
FW     720
DF    1200
MD    1257
dtype: int64

We could now go on and check the distribution of the strong foot etc. However, let's instead look into a great package for initial exploratory data analysis: [pandas-profiling](https://github.com/pandas-profiling/pandas-profiling) 

Pandas-profiling creates a small html with the most important statistics for exploratory data analysis such as:
- Uniqueness
- Missing values
- Histograms
- Min / max / mean
- Column type

The html can either be saved in a html-file or directly be shown in a Jupyter notebook.


In [7]:
profile = ProfileReport(df_players, title='Player Report', minimal=True)
profile.to_notebook_iframe()

c:\programs_clemens\anaconda3\envs\soccer_analytics\lib\site-packages\IPython\core\display.py:701: UserWarning:

Consider using IPython.display.IFrame instead



That's great, isn't it?! :-) That way we can directly get what we computed above with only one line of code (& it looks way coooler)

<a id="team_data"></a>

## Team data

Let's now look into the team data. We first read all teams of the German Bundesliga

In [8]:
df_teams = io.read_team_data("germany")

In [9]:
df_teams

position  teamId             teamName  matches  goals  concededGoals  \
4          1    2444       Bayern München       34     92             28   
6          2    2449           Schalke 04       34     53             37   
12         3    2482           Hoffenheim       34     66             48   
3          4    2447    Borussia Dortmund       34     64             47   
2          5    2446     Bayer Leverkusen       34     58             44   
17         6    2975           RB Leipzig       34     57             53   
5          7    2445            Stuttgart       34     36             36   
15         8    2462  Eintracht Frankfurt       34     45             45   
0          9    2454  Borussia M'gladbach       34     47             52   
8         10    2457           Hertha BSC       34     43             46   
1         11    2443        Werder Bremen       34     37             40   
13        12    2481             Augsburg       34     43             46   
7         13    2455          Hannover 96       34     44             54   
14        14    2460             Mainz 05       34     38             52   
11        15    2453             Freiburg       34     32             56   
9         16    2451            Wolfsburg       34     36             48   
10        17    2450         Hamburger SV       34     29             53   
16        18    2463                 Köln       34     35             70   

    goalsDiff  points  
4          64      84  
6          16      63  
12         18      55  
3          17      55  
2          14      55  
17          4      53  
5           0      51  
15          0      49  
0          -5      47  
8          -3      43  
1          -3      42  
13         -3      41  
7         -10      39  
14        -14      36  
11        -24      36  
9         -12      33  
10        -24      31  
16        -35      22

Again, I guess that every column is self-explaining. Notice that each team has both a *teamId* and a *teamName*. We will need to use those two columns quite regularly in the upcoming notebooks.

<a id="match_data"></a>

## Match data


In [10]:
df_matches = io.read_match_data("germany")

In [11]:
df_matches.head()

matchId  gameweek             dateutc  teamId  side  score  oppTeamId  \
305  2516739         1 2017-08-18 18:30:00    2444  home      3       2446   
305  2516739         1 2017-08-18 18:30:00    2446  away      1       2444   
300  2516740         1 2017-08-19 13:30:00    2482  home      1       2443   
300  2516740         1 2017-08-19 13:30:00    2443  away      0       2482   
301  2516741         1 2017-08-19 13:30:00    2457  home      2       2445   

     oppScore  points  scoreDiff  
305         1       3          2  
305         3       0         -2  
300         0       3          1  
300         1       0         -1  
301         0       3          2

Ok, the match data obviously contains the matches of the Bundesliga. 

But why does e.g. the first matchId appear in both the first and the second row? The idea is to split each match into two perspectives and look at it from both the home team's and the away team's perspective. 

Let's take the first match: 

1. In the first line it says that Bayern (2444) played at home against Leverkusen (2446), scored 3 goals itself and the opponent scored 1 goal, e.g. Bayern won 3-1 against Leverkusen. This is Bayern's perspective.
2. In the second line it says that Leverkusen (2446) played away against Bayern (2444), scored 1 goal itself and the opponent scored 3 goals, e.g. Leverkusen lost 1-3 at Bayern. This Leverkusen's perspective.

Makes sense, doesn't it? So why do we do that? This way we can very easily extract e.g. the total number of goals Bayern scored in the whole season:

In [12]:
print(f"Bayern scored in total {df_matches[df_matches['teamId'] == 2444]['score'].sum()} goals")

Bayern scored in total 92 goals


Great, that matches exactly the 92 goals that we saw in the *team_data* above.

To make sure there is nothing weird in the data, let's run pandas-profiling again

In [13]:
profile = ProfileReport(df_matches, title='Match Report', minimal=True)
profile.to_notebook_iframe()

c:\programs_clemens\anaconda3\envs\soccer_analytics\lib\site-packages\IPython\core\display.py:701: UserWarning:

Consider using IPython.display.IFrame instead



Looks good to me! So let's continue with the formation data

<a id="formation_data"></a>

## Formation data


The formation data contains the line-up as well as the substitutes for all matches.

In [14]:
df_formations = io.read_formation_data("germany")

In [15]:
df_formations.head()

playerId  lineup  matchId  teamId  substituteIn  substituteOut  \
0    209091       1  2517036    2444             0              1   
1     14732       1  2517036    2444             0              1   
2     14726       1  2517036    2444             0              1   
3     14736       1  2517036    2444             0              0   
4     14817       1  2517036    2444             0              0   

   minuteStart  minuteEnd  minutesPlayed  
0          0.0       46.0           46.0  
1          0.0       68.0           68.0  
2          0.0       71.0           71.0  
3          0.0       90.0           90.0  
4          0.0       90.0           90.0

Let's look at Bayern (2444) in their first match (2517036):

In [16]:
df_formations[(df_formations["matchId"] == 2517036) & (df_formations["teamId"] == 2444)]

playerId  lineup  matchId  teamId  substituteIn  substituteOut  \
0     209091       1  2517036    2444             0              1   
1      14732       1  2517036    2444             0              1   
2      14726       1  2517036    2444             0              1   
3      14736       1  2517036    2444             0              0   
4      14817       1  2517036    2444             0              0   
5     134383       1  2517036    2444             0              0   
6       3345       1  2517036    2444             0              0   
7     224593       1  2517036    2444             0              0   
8      14718       1  2517036    2444             0              0   
9      70096       1  2517036    2444             0              0   
10     14795       1  2517036    2444             0              0   
11     15797       0  2517036    2444             0              0   
12    471945       0  2517036    2444             0              0   
13    391442       0  2517036    2444             0              0   
14    433100       0  2517036    2444             0              0   
15     15811       0  2517036    2444             1              0   
16      3416       0  2517036    2444             1              0   
17     15085       0  2517036    2444             1              0   

    minuteStart  minuteEnd  minutesPlayed  
0           0.0       46.0           46.0  
1           0.0       68.0           68.0  
2           0.0       71.0           71.0  
3           0.0       90.0           90.0  
4           0.0       90.0           90.0  
5           0.0       90.0           90.0  
6           0.0       90.0           90.0  
7           0.0       90.0           90.0  
8           0.0       90.0           90.0  
9           0.0       90.0           90.0  
10          0.0       90.0           90.0  
11          0.0        0.0            0.0  
12          0.0        0.0            0.0  
13          0.0        0.0            0.0  
14          0.0        0.0            0.0  
15         71.0       90.0           19.0  
16         46.0       90.0           44.0  
17         68.0       90.0           22.0

There are 4 different groups of players:
1. Those that started the match (*lineup* = 1) and did not get substituted (*substituteOut* = 0)
2. Those that started the match (*lineup* = 1) and did get substituted (*substituteOut* = 1)
3. Those that started on the bench (*lineup* = 0) but did get substituted in (*substituteIn* = 1)
4. Those that started on the bench (*lineup* = 0) but did not get substituted in (*substituteIn* = 0)

Let's run pandas-profiling again. Make sure to also click on the "Toggle details" button for *minuteStart*:
- Most substitutions happen between minute 70 and 80
- The minute with the most substitutions is the 46th (not surprisingly :-))
- The first substitutions already happened after 8 minutes

In [17]:
profile = ProfileReport(df_formations, title='Formation Report', minimal=True)
profile.to_notebook_iframe()

c:\programs_clemens\anaconda3\envs\soccer_analytics\lib\site-packages\IPython\core\display.py:701: UserWarning:

Consider using IPython.display.IFrame instead



<a id="event_data"></a>

## Event data


The by far most interesting data is the event data. This is the one that we will use most often and also the one that is the least self-explanatory. 

In [8]:
df_events = io.read_event_data("germany")

In [20]:
print(f"There are in total {len(df_events)} rows")

There are in total 519407 rows


In [21]:
df_events.head()

id  matchId matchPeriod   eventSec eventName subEventName  teamId  \
0  179896442  2516739          1H   2.409746      Pass  Simple pass    2446   
1  179896443  2516739          1H   2.506082      Pass  Simple pass    2446   
2  179896444  2516739          1H   6.946706      Pass  Simple pass    2446   
3  179896445  2516739          1H  10.786491      Pass  Simple pass    2446   
4  179896446  2516739          1H  12.684514      Pass  Simple pass    2446   

   posBeforeXMeters  posBeforeYMeters  posAfterXMeters  posAfterYMeters  \
0              52.5             34.00            52.50            32.64   
1              52.5             32.64            23.10            14.96   
2              23.1             14.96             6.30            31.28   
3               6.3             31.28            21.00             6.80   
4              21.0              6.80            28.35             2.72   

   playerId    playerName playerPosition playerStrongFoot  teamPossession  \
0     15231    K. Volland             FW             left            2446   
1     14786  K. Bellarabi             MD            right            2446   
2     14803     S. Bender             DF            right            2446   
3     14768       B. Leno             GK            right            2446   
4     14803     S. Bender             DF            right            2446   

   homeTeamId  awayTeamId  accurate  notAccurate  goal  ownGoal  assist  \
0        2444        2446         1            0     0        0       0   
1        2444        2446         1            0     0        0       0   
2        2444        2446         1            0     0        0       0   
3        2444        2446         1            0     0        0       0   
4        2444        2446         1            0     0        0       0   

   keyPass  counterAttack  leftFoot  rightFoot  head/body  direct  indirect  \
0        0              0         0          0          0       0         0   
1        0              0         0          0          0       0         0   
2        0              0         0          0          0       0         0   
3        0              0         0          0          0       0         0   
4        0              0         0          0          0       0         0   

   dangerousBallLost  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

Let's start with profiling before we go into more detail below. Notice that this can take a little while as there is quite a lot of data

In [22]:
profile = ProfileReport(df_events, title='Event Report', minimal=True)
profile.to_notebook_iframe()

c:\programs_clemens\anaconda3\envs\soccer_analytics\lib\site-packages\IPython\core\display.py:701: UserWarning: Consider using IPython.display.IFrame instead
  warnings.warn("Consider using IPython.display.IFrame instead")


I guess that already gives you a rather good idea of the event data. However, let's go over some of the columns in a little bit more detail

### Time of the event

The *eventSec* is obviously the second when the event took place. Notice, however, that the second is always w.r.t the half-time that is stored in the *matchPeriod* column. When looking at subsequent events you should therefore always make sure you consider those 2 columns together.

### Position of the event

There are 4 position columns:
- *posBeforeXMeters*
- *posAfterXMeters*
- *posBeforeYMeters*
- *posAfterYMeters*

As the column name suggests all of the values are measured in meters assuming a field size of 105x68 meters. The *X* values are measured along the length of field and therefore take values between 0 and 105, whereas the *Y* values are measured along the width of the field and take values between 0 and 68 meters.

#### Position depends on team performing event

Imagine a soccer field: Values of *X* and *Y* are always measured from the top left of the field from the perspective of the team performing the event. 

Examples:
1. If team A has a penalty, this is taken place from the position (94,34): 94 as 105-11 = 94 and 34 as 68 / 2 = 34
2. The goalie saves the ball and has it in his hands. He is then in position (0,34): Remember that it is always from the perspective of the team performing the event, therefore *X*=0 from team B's perspective
3. The goalie throws the ball to his left full-back being at e.g. (15, 12): *Y* values < 34 are always the left-hand side of the field
4. The full-back immediately loses the ball to an attacker who is then in (90, 56): 105-15 = 90 and 68-12 = 56

I hope this makes sense? If not yet, it will definitely make sense after we go into the tutorials and visualize some of these events.

#### Position before vs. after

As the name suggests, this is the position before and after the event happened. E.g. consider the following pass:

In [23]:
df_events[df_events["id"]==179896442]

id  matchId matchPeriod  eventSec eventName subEventName  teamId  \
0  179896442  2516739          1H  2.409746      Pass  Simple pass    2446   

   posBeforeXMeters  posBeforeYMeters  posAfterXMeters  posAfterYMeters  \
0              52.5              34.0             52.5            32.64   

   playerId  playerName playerPosition playerStrongFoot  teamPossession  \
0     15231  K. Volland             FW             left            2446   

   homeTeamId  awayTeamId  accurate  notAccurate  goal  ownGoal  assist  \
0        2444        2446         1            0     0        0       0   

   keyPass  counterAttack  leftFoot  rightFoot  head/body  direct  indirect  \
0        0              0         0          0          0       0         0   

   dangerousBallLost  
0                  0

In [24]:
df_events[df_events["id"]==179896443][["eventName", "posBeforeXMeters", "posBeforeYMeters", 
                                       "posAfterXMeters", "posAfterYMeters"]]

eventName  posBeforeXMeters  posBeforeYMeters  posAfterXMeters  \
1      Pass              52.5             32.64             23.1   

   posAfterYMeters  
1            14.96

This was a pass going from (52.5, 32.64) to (23.1, 14.96), e.g. from the very middle of the field back to the right full-back

#### Event type

We have seen in the profiling that there are 10 different *eventName* and 36 different *subEventName* with the following number of appearances:

In [25]:
df_events.groupby("eventName").size().sort_values(ascending=False)

eventName
Pass                       261462
Duel                       144183
Others on the ball          40461
Free Kick                   31086
Interruption                21716
Foul                         8656
Shot                         6898
Save attempt                 2811
Offside                      1219
Goalkeeper leaving line       915
dtype: int64

In [26]:
df_events.groupby("subEventName").size().sort_values(ascending=False)

subEventName
Simple pass                198730
Ground attacking duel       44342
Ground defending duel       44165
Air duel                    29448
Touch                       27700
Ground loose ball duel      26228
Ball out of the field       21588
High pass                   20861
Head pass                   17085
Throw in                    13682
Cross                        9031
Clearance                    8801
Launch                       8714
Foul                         8115
Free Kick                    7748
Shot                         6898
Goal kick                    4847
Smart pass                   4623
Acceleration                 3960
Corner                       2891
Hand pass                    2418
Reflexes                     1815
Free kick cross              1526
Offside                      1219
Save attempt                  996
Goalkeeper leaving line       915
Hand foul                     311
Free kick shot                299
Whistle                       128
P

Instead of me trying to explain what all of those are, I'll give you a link to the wyscout website where everything is described in great detail: [Event description](https://footballdata.wyscout.com/events-manual/)

#### Binary columns

There are a couple of binary columns such as *goal* and *ownGoal* or *leftFoot*. We will not go into great detail here, but <font color=red>**be careful when using those**</font>. Reason is that they might only be filled (or worse, filled correctly) for certain event types.

Let's first look at the number of total goals

In [27]:
goals_event_data = df_events["goal"].sum() + df_events["ownGoal"].sum()
goals_match_data = df_matches["score"].sum() + df_matches["oppScore"].sum()
print(f"Total number of goals in event data: {goals_event_data}")
print(f"Total number of goals in match data: {goals_match_data}")

Total number of goals in event data: 1711
Total number of goals in match data: 1710


Ok, the event data has one goal more than the match data. I guess that is something we can live with... 

Let's now assume we want to do an analysis on the whether a shot or pass was done with the left foot / right foot / head. 

We start with the shots:

In [28]:
df_shots = df_events[df_events["eventName"] == "Shot"][["eventSec", "eventName", "leftFoot", "rightFoot", "head/body"]]

In [29]:
df_shots.head()

eventSec eventName  leftFoot  rightFoot  head/body
104  247.703507      Shot         0          1          0
178  529.393731      Shot         0          0          1
216  668.234340      Shot         0          1          0
220  672.925920      Shot         0          1          0
313  949.131592      Shot         0          1          0

Ok, it looks like all shots actually have an indication regarding the part of the body they were taken with. Let's check this more consistently...

In [30]:
df_shots["bodyPartAvailable"] = 1*(df_shots["leftFoot"] + df_shots["rightFoot"] + df_shots["head/body"])
print(f"Body part available for {df_shots['bodyPartAvailable'].sum() / len(df_shots) * 100 : .1f}% of the shots")

Body part available for  100.0% of the shots


Amazing, we do seem to have the body for every shot that was taken. Let's now look at passes

In [31]:
df_passes = df_events[df_events["eventName"] == "Pass"][["eventSec", "eventName", "leftFoot", "rightFoot", "head/body"]]
df_passes["bodyPartAvailable"] = 1*(df_passes["leftFoot"] + df_passes["rightFoot"] + df_passes["head/body"])
print(f"Body part available for {df_passes['bodyPartAvailable'].sum() / len(df_passes) * 100 : .1f}% of the passes")

Body part available for  3.5% of the passes


Hm, the body part only seems to be available for 3.5% of the passes. This is probably nothing we want to do an analysis on... So, again, be very careful before you dive into an analysis!

## Summary

You did it, you just went through the first (admittedly rather dry) notebook! Great job!

In this notebook you learned about:
- the different data tables that are available on what each of them contains
- how to use the pandas-profiling package
- that you need to be careful with the data before jumping into analysis

In the next notebook we are going to look into goal kicks made by the different teams and learn about some cool visualization techniques on the way. Have fun! 